File: src/main/java/com/tw/pos/parse/DiscountParse.java
package com.tw.pos.parse;


import com.tw.pos.discount.Discount;
import com.tw.pos.discount.DiscountPromotion;
import com.tw.pos.discount.SecondHalf;
import com.tw.pos.entity.Goods;
import com.tw.pos.entity.GoodsRepository;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

public class DiscountParse {

    private static final Pattern DISCOUNT_PATTERN = Pattern.compile("^\\w+:\\d+$");
    private static final Pattern SECOND_HALF_PATTERN = Pattern.compile("^\\w+$");

    public List<DiscountPromotion> parseDiscount(List<String> discountInfoStrList) {
        List<DiscountPromotion> discountPromotions = new ArrayList<>();
        discountInfoStrList.forEach(discountInfo->{
            if (!DISCOUNT_PATTERN.matcher(discountInfo).matches() && !SECOND_HALF_PATTERN.matcher(discountInfo).matches()){
                throw new IllegalArgumentException("invalid discountInfo input");
            }

            if (DISCOUNT_PATTERN.matcher(discountInfo).matches()){
                String[] discountInfoStr = discountInfo.split(":");
                Goods goods = GoodsRepository.getGoodsByBarcode(discountInfoStr[0]);
                DiscountPromotion discount = new Discount(goods, Double.valueOf(discountInfoStr[1]));
                discountPromotions.add(discount);
            }

            if (SECOND_HALF_PATTERN.matcher(discountInfo).matches()){
                Goods goods = GoodsRepository.getGoodsByBarcode(discountInfo.trim());
                DiscountPromotion secondHalf = new SecondHalf(goods);
                discountPromotions.add(secondHalf);
            }
        });
        return discountPromotions;
    }
}

File: js/parse/discountParse.js

var DiscountParse = function(){
    this.discountRxp = new RegExp('^\\w+:\\d+$');
    this.secondHalfRxp = new RegExp('^\\w+$');
};

DiscountParse.prototype.parse = function(discountInfoStrList){
    var discountMap = {};
    for (var i = 0, discountInfoStr; discountInfoStr = discountInfoStrList[i++];){
        if (!this.discountRxp.test(discountInfoStr) && !this.secondHalfRxp.test(discountInfoStr)){
            throw new Error('invalid discountInfo input');
        }

        if (this.discountRxp.test(discountInfoStr)){
            var discountInfoList = discountInfoStr.split(':');

            var barcode = discountInfoList[0];
            var discount = new DiscountPromotion(parseFloat(discountInfoList[1]));
            this.addDiscount(discountMap, barcode, discount);
        }

        if (this.secondHalfRxp.test(discountInfoStr)){
            var barcode = discountInfoStr.trim();
            var discount = new SecondHalf();
            this.addDiscount(discountMap, barcode, discount);
        }
    }
    return discountMap;
};

DiscountParse.prototype.addDiscount= function(discountMap, barcode, discount){
    var discountList = discountMap[barcode];
    if (discountList == null || discountList.size == 0){
        discountList = [];
    }
    if (discount instanceof DiscountPromotion){
        discountList.splice(0,0,discount);
    }else {
        discountList.push(discount);
    }
    discountMap[barcode] = discountList;
};

